### ISSUE TITLE
Opengl on windows 10 with shared library

### ISSUE THREAD

**Issue by LucaRitz** [fltk/fltk#136]
Hello there

I've encountered a problem using the shared-build version of fltk "release-1.3.5".
Maybe, I'm doing something really stupid or it is just a bug.

I've added the dependency on fltk the following way (CMake):
```
include(FetchContent)
set(OPTION_BUILD_PDF_DOCUMENTATION OFF CACHE INTERNAL "")
set(OPTION_BUILD_SHARED_LIBS ON CACHE INTERNAL "")
set(OPTION_USE_SYSTEM_ZLIB OFF CACHE INTERNAL "")
set(OPTION_USE_SYSTEM_LIBPNG OFF CACHE INTERNAL "")
set(OPTION_USE_SYSTEM_LIBJPEG OFF CACHE INTERNAL "")

FetchContent_Declare(
        fltk
        GIT_REPOSITORY https://github.com/fltk/fltk
        GIT_TAG        release-1.3.5
)
FetchContent_MakeAvailable(fltk)
unset(OPTION_BUILD_PDF_DOCUMENTATION)
unset(OPTION_BUILD_SHARED_LIBS)
unset(OPTION_USE_SYSTEM_ZLIB)
unset(OPTION_USE_SYSTEM_LIBPNG)
unset(OPTION_USE_SYSTEM_LIBJPEG)

target_compile_definitions(fltk_SHARED PUBLIC -DFL_DLL)
target_compile_definitions(fltk_gl_SHARED PUBLIC -DFL_DLL)
```

When I execute the following code, which I've found on [Opengl-Example](https://research.cs.wisc.edu/graphics/Courses/638-f1999/gl_survival.htm)  and changed a bit, it shows me a grey window with a button. The draw-function of the Fl_Gl_Window is never called so the window is not displayed.
Link against: fltk_SHARED fltk_gl_SHARED

```
#include <glad/glad.h>
#include <FL/Fl_Gl_Window.h>
#include <Fl/Fl.h>
#include <windows.h>
#include <GL/gl.h>
#include <Fl/Fl_Double_Window.h>
#include <Fl/Fl_Button.h>
#include <iostream>

// make a Gl window that draws something:
class MyGlWindow : public Fl_Gl_Window {
public:
    float r,g,b;
    MyGlWindow(int x, int y, int w, int h) :
            Fl_Gl_Window(x,y,w,h,"My GL Window")
    {
        r = g = 1;
        b = 0;
    }
private:
    void draw() {	// the draw method must be private
        std::cout << "Start GL" << std::endl;
        make_current();

        static bool first = true;
        if (first) {
            std::cout << "Set viewport" << std::endl;
            // make_current();
            if (!gladLoadGL()) {
                std::cerr << "Failed to initialize OpenGL context" << std::endl;
            } else {

            }
            glViewport(0, 0, 10, 10);
            first = false;
        }
        std::cout << "Viewport set" << std::endl;
        glClearColor(0.2f, 0.3f, 0.3f, 1.0f);
        glClear(GL_COLOR_BUFFER_BIT);
        glLoadIdentity();
        glFlush();
        glEnd();
        std::cout << "Ok" << std::endl;
    };
};

// a simple callback to do something
void changeColor(Fl_Widget*
button
, MyGlWindow* myWind)
{
    myWind->r = 1 - myWind->r;
    myWind->g = 1 - myWind->g;
    myWind->b = 1 - myWind->b;
    myWind->damage(1);
}

// the main routine makes the window, and then runs an even loop
// until the window is closed
int main()
{
    Fl_Double_Window* wind = new Fl_Double_Window(100,100,400,300,"GL Sample");

    wind->begin();		// put widgets inside of the window
    MyGlWindow* gl = new MyGlWindow(10,10,280,280);
    Fl_Button* bt = new Fl_Button(300,10,70,25,"Color");
    bt->callback((Fl_Callback*) changeColor, gl);
    wind->end();

    wind->show();	// this actually opens the window

    Fl::run();
    delete wind;

    return 1;
}
```

Output:
![opengl_shared](https://user-images.githubusercontent.com/36641286/93196397-aa002c00-f74a-11ea-9d07-1b2c332c50ba.PNG)

When I change "OPTION_BUILD_SHARED_LIBS " to OFF or just link against "fltk fltk_gl" it works, the window gets drawn.
Output:
![opengl_static](https://user-images.githubusercontent.com/36641286/93196418-b2586700-f74a-11ea-9a6e-100ec9e2aae4.PNG)

Additional information:
MSVC: 14.24.28314
CMake-Version: 3.18.2
Architecture: x64
OS: Windows 10 Pro (Version	10.0.18362 Build 18362)
IDE: CLion 2019.3

Thank you for your help, maybe I'm just missing something.

**Comment by Albrecht-S** [comment]
Thanks for the very good report including the demo program, and sorry for this **known issue**.

The shared library build of FLTK 1.3.5 with CMake using MSVC is currently broken, i.e. shared libs don't import/export the correct symbols. The only shared library that works is fltk_SHARED alone; you can't link with fltk_images or fltk_gl. It is correct that the non-shared build works as you also noticed. The bundled MSVC project should also build correctly (only one shared library), but this would lack the CMake import files.

This issue is (should be) fixed in FLTK 1.4, but it's not yet clear if this can or will be backported to 1.3 and whether there will be a release 1.3.6 in the future.

Please pull the `master` branch from Git, build the shared libs and confirm that it works for you. This would be very helpful for our development. It's all API compatible.

Note that some of the `option` names may be different in 1.4 and the library _**file names**_ are different, but you should be able to build and run your program just fine. The CMake library names `fltk_SHARED fltk_gl_SHARED` are still the same though so your program should hopefully build just fine.

Any feedback would be appreciated. TIA.

**Comment by LucaRitz** [comment]
Thank you very much for your quick response.
I've tried to build the master branch (origin/master), but I get the following error (I'm sorry, it's in german ;) :
```
[ 19%] Built target fltk_z_SHARED
NMAKE : fatal error U1073: "_deps\fltk-build\lib\libfltk_zd.lib" konnte nicht erstellt werden
Stop.
NMAKE : fatal error U1077: ""C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\bin\HostX64\x64\nmake.exe"": Rückgabe-Code "0x2"
Stop.
NMAKE : fatal error U1077: ""C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\bin\HostX64\x64\nmake.exe"": Rückgabe-Code "0x2"
Stop.
NMAKE : fatal error U1077: ""C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\bin\HostX64\x64\nmake.exe"": Rückgabe-Code "0x2"
Stop.
```
The file "libfltk_zd.lib" cannot be created.

**Comment by Albrecht-S** [comment]
German is not a problem for **me**. ;-)

Sorry, I can't help with these MS (VS related) error messages. I'm not sure why you get `NMAKE` error messages at all. How did you (try to) build FLTK? `cmake --build` maybe? If I'm using VS I'm usually using the IDE, but ... didn't you write you're using CLion? That's something I never used, so you may be on your own (although someone else might be able to help).

The only idea to help from me would be to do a fresh checkout in a new build directory in case you didn't clean up everything correctly. HTH.

**Comment by LucaRitz** [comment]
Ok, that's nice ;)

I've turned the following options to ON:
OPTION_USE_SYSTEM_ZLIB
OPTION_USE_SYSTEM_LIBPNG
OPTION_USE_SYSTEM_LIBJPEG
just to test if it works now and it seems to be fine, I get my expected result.

Yes you're right, I'm using CLion but I have deleted the whole build and source folders, so there shouldn't be something cached used.
I think CLion does some magic behind the scenes with CMake. I try to checkout and build it with cmake only.

Thank you very much for your help. For me, this "bug" is fixed in 1.4. The other I mentioned is something new, which does not really belong here and is maybe a bug in CLion, which I try to find out.
Do you agree with that?

**Comment by Albrecht-S** [comment]
Yes, I agree. The mentioned options _should not_ have any impact on the build, but maybe there's an incompatibility of the bundled image libs with something else.

Thanks for confirmation that it works in 1.4

Ready to close? Feel free to close it yourself...
